rfcs/0042-config-option.md

```diff
@@ -0,0 +1,301 @@
+
+# Motivation
+[motivation]: #motivation
+
+NixOS commonly has 2 models of specifying configuration for programs, each with their own set of problems. This RFC aims to solve all of them.
+
+## Single option for every setting
+
+Having a single option for every setting in the configuration file, this often gets combined with an `extraConfig` option to provide greater flexibility. Problems:
+
+- Coupling to upstream
+  - When upstream adds or removes settings, the NixOS module needs to be updated to reflect that.
+    - Upstream adds a setting: If the module has an `extraConfig` option people might set the new setting there. But if we ever add it as a NixOS option, we'll have trouble merging the values together with what the user already specified in `extraConfig`
+    - Upstream removes a setting (backwards incompatible): The NixOS module is straight up broken in nixpkgs until somebody fixes it, end users can't fix it themselves (unless the module provides a `configFile` option which can override the generated one). The same can also happen when the user uses overlays or `disabledModules` to cause a module/package version mismatch.
+  - The documentation of the upstream settings needs to be copied over to the NixOS options, which means it might get out of sync with changes
+  - The upstream defaults are being copied to the NixOS modules, so we need to also update the defaults whenever upstream changes them. This can be solved by using `nullOr` to allow for a `null` value to indicate that the upstream default shall be used, but that's not very nice.
+- Option bloat: <s>NixOS evaluation time is still linear in the number of *available* options for all users, even though everybody only uses a fraction of them. This means that when a module adds 100 new options, this leads to a direct increase in evaluation time for every `nixos-rebuild switch` of everybody using NixOS.</s> With high confidence debunked in [#57477](https://github.com/NixOS/nixpkgs/issues/57477).
+- Timeconsuming to implement, tedious to review and hard to maintain
+  - It takes a non-zero amount of time to write all the wanted options out into a NixOS module
+  - Reviewing is tedious because people need to make sure types are correct, descriptions are fitting, defaults are acceptable, for every option added. Due to the size and repetitiveness, people are also less willing to thoroughly review the code.
+  - The bigger the NixOS module is the harder it is to maintain, and the less people want to actually maintain it. 
+- Responsibility for backwards compatibility: By adding such options, we obligate ourselves to handle backwards incompatibilites on our side. We will have to support these options for a long time and can't remove them without at least one person being annoyed about it.
```

Comment:
It strikes me that nearly all of the problems listed above are related to maintenance burden. The primary technical arguments I see are

- the issue of merging the module config with the user provided `extraConfig` and
- the issue of the module and package going out of sync due to module and package version mismatch.

In Home Manager we have in a few modules attempted to mitigate the first issue by allowing a "structured `extraConfig`" where this is possible, typically by letting it have an `attrsOf *` type. For example, [`programs.notmuch.extraConfig`](https://github.com/rycee/home-manager/blob/f8b03f5750679d6686eefa60d5de8cfd932b8997/modules/programs/notmuch.nix#L91-L97) and [`programs.git.extraConfig`](https://github.com/rycee/home-manager/blob/f8b03f5750679d6686eefa60d5de8cfd932b8997/modules/programs/git.nix#L115-L122), where the latter for backwards compatibility allow _either_ a string or an attribute set as value. I haven't received any complaints about this and I think having a structured `extraConfig` option is generally appreciated by the users of these modules.

In Home Manager we haven't consciously attempted anything related to the second issue but I think the scheme described in this RFC will mostly resolve this. Even if upstream suddenly changes the configuration format to make an option key illegal it would be possible for the module user to `mkForce` the entire config attribute set (although I suspect this is not practical for larger modules).